src/librustc_typeck/astconv.rs

```diff
@@ -1021,7 +1029,7 @@ impl<'o, 'gcx: 'tcx, 'tcx> dyn AstConv<'gcx, 'tcx>+'o {
             return tcx.types.err;
         }
 
-        // Use a BTreeSet to keep output in a more consistent order.
+        // Use a `BTreeSet` to keep output in a more consistent order.
         let mut associated_types = BTreeSet::default();
 
         for tr in traits::supertraits(tcx, principal) {
```

Comment:
Ok so I would advise merging the code which extends `projection_bounds` with this loop, something like:
```rust
for predicate in traits::elaborate_trait_ref(tcx, principal).into_iter() {
    match predicate {
        ty::TraitPredicate(tr) => {
            associated_types.extend(tcx.associated_items(tr.def_id())
                .filter(|item| item.kind == ty::AssociatedKind::Type)
                .map(|item| item.def_id));
        }
        ty::ProjectionPredicate(projection) => projection_bounds.push((projection, DUMMY_SP)),
        _ => continue,
    }
}
```
also, turns out we don't need any span information in the end, so `DUMMY_SP` is fine and we don't need to tweak `elaborate_trait_ref`.